src/main/java/pl/arturzgodka/databaseutils/CharacterDao.java

```diff
@@ -0,0 +1,150 @@
+package pl.arturzgodka.databaseutils;
+
+import jakarta.persistence.criteria.CriteriaBuilder;
+import jakarta.persistence.criteria.CriteriaQuery;
+import jakarta.persistence.criteria.Root;
+import org.hibernate.Session;
+import org.hibernate.SessionFactory;
+import org.hibernate.Transaction;
+import pl.arturzgodka.datamodel.*;
+
+import java.util.List;
+import java.util.Map;
+
+public class CharacterDao {
+
+    private final SessionFactory sessionFactory = UserSessionFactory.getCustomUserSessionFactory();
+    public void saveCharacter(CharacterDataModel characterDataModel) {
+        Session session = sessionFactory.openSession();
+        Transaction t = session.beginTransaction();
+        session.merge(characterDataModel);
+        t.commit();
+        session.close();
+    }
+
+    public void deleteCharacter(CharacterDataModel characterDataModel) {
+        Session session = sessionFactory.openSession();
+        Transaction transaction = session.beginTransaction();
+        session.remove(characterDataModel);
+        transaction.commit();
+        session.close();
+    }
+
+    public void deleteCharacter(int id) {
+        Session session = sessionFactory.openSession();
+        Transaction transaction = session.beginTransaction();
+        CriteriaBuilder cb = session.getCriteriaBuilder();
+        CriteriaQuery<CharacterDataModel> userQuery = cb.createQuery(CharacterDataModel.class);
+        Root<CharacterDataModel> root = userQuery.from(CharacterDataModel.class);
+        userQuery.select(root).where(cb.equal(root.get("id"), id));
+        session.remove(session.createQuery(userQuery).getSingleResult());
+        transaction.commit();
+        session.close();
+    }
+
+    private boolean characterExists(CharacterDataModel characterDataModel) {
+        return findCharacterById(characterDataModel.getId()) != null;
+    }
+
+    public CharacterDataModel findCharacterById(int id) {
```

Comment:
should this be public?

Reply:
I made it public for now to use in assertions in test class.